docs/architecture/adr-038-state-listening.md

```diff
@@ -0,0 +1,295 @@
+# ADR 038: KVStore state listening
+
+## Changelog
+
+- 11/23/2020: Initial draft
+
+## Status
+
+Proposed
+
+## Abstract
+
+This ADR defines a set of changes to enable listening to state changes of individual KVStores and exposing these data to consumers.
+
+## Context
+
+Currently, KVStore data can be remotely accessed through [Queries](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/messages-and-queries.md#queries) which proceed through Tendermint and the ABCI.
+In addition to these request/response queries, it would be beneficial to have a means of listening to state changes as they occur in real time.
+
+## Decision
+
+We will modify the `MultiStore` interface and its concrete (`rootmulti` and `cachemulti`) implementations and introduce a new `listenkv.Store` to allow listening to state changes in underlying KVStores.
+We will also introduce two approaches for exposing the data to external consumers: writing to files and writing to a gRPC stream.
+
+### Listening interface
+In a new file- `store/types/listening.go`- we will create a `WriteListener` interface for streaming out state changes from a KVStore.
+
+```go
+// WriteListener interface for writing data out from a listenkv.Store
```

Comment:
From reading and understanding this, `WriteListener` is more of a streaming proxy from a producer (module KVStore) to a consumer (essentially any `io.Writer`), there inherently does not need to be any writing on the consumer-side, although obviously there will be. I would rephrase this godoc.

Reply:
Absolutely, "streaming data out" would be more accurate.